source/4_tools/regina/src/regina.cpp

```diff
@@ -1,54 +1,865 @@
+            });
+
+        return it != rsl::cend(g_regina_ctx.frame_resource_data)
+            ? it->pass_commited_resource
+            : renderer::ResourceSlot::make_invalid();
+    }
+    //-------------------------------------------------------------------------
+    renderer::ResourceSlot get_active_object_constant_buffer_for_frame(s32 frame, s32 idx)
+    {
+        auto it = rsl::find_if(rsl::cbegin(g_regina_ctx.frame_resource_data), rsl::cend(g_regina_ctx.frame_resource_data), 
+            [frame](const FrameData& data)
+        {
+                return frame == data.frame.slot_id();
+        });
+
+        return it != rsl::cend(g_regina_ctx.frame_resource_data)
+            ? it->object_constant_buffer[idx]
+            : renderer::ResourceSlot::make_invalid();
+    }
+    //-------------------------------------------------------------------------
+    renderer::ResourceSlot get_active_pass_constant_buffer_for_frame(s32 frame)
+    {
+        auto it = rsl::find_if(rsl::cbegin(g_regina_ctx.frame_resource_data), rsl::cend(g_regina_ctx.frame_resource_data),
+            [frame](const FrameData& data)
+            {
+                return frame == data.frame.slot_id();
+            });
+
+        return it != rsl::cend(g_regina_ctx.frame_resource_data)
+            ? it->pass_constant_buffer
+            : renderer::ResourceSlot::make_invalid();
+    }
+
+    //-------------------------------------------------------------------------
+    renderer::commands::CompileShaderCommandDesc create_compile_shader_parameters(const rsl::small_stack_string& shaderName, renderer::ShaderType shaderType, rsl::string_view filePath)
+    {
+        renderer::commands::CompileShaderCommandDesc compile_shader_command_desc;
+
+        switch (shaderType)
+        {
+        case renderer::ShaderType::VERTEX:
+            compile_shader_command_desc.shader_entry_point = rsl::tiny_stack_string("VS");
+            compile_shader_command_desc.shader_feature_target = rsl::tiny_stack_string("vs_5_1");
+            break;
+        case renderer::ShaderType::PIXEL:
+            compile_shader_command_desc.shader_entry_point = rsl::tiny_stack_string("PS");
+            compile_shader_command_desc.shader_feature_target = rsl::tiny_stack_string("ps_5_1");
+            break;
+        }
+
+        compile_shader_command_desc.shader_code = vfs::open_read(filePath);
+        compile_shader_command_desc.shader_name = shaderName;
+        compile_shader_command_desc.shader_type = shaderType;
+
+        return compile_shader_command_desc;
+    }
+
+    //-------------------------------------------------------------------------
+    template<typename T>
+    renderer::commands::CreateBufferCommandDesc create_buffer_parameters(T* data, s32 num)
+    {
+        renderer::commands::CreateBufferCommandDesc create_buffer_command_desc;
+
+        create_buffer_command_desc.buffer_data = memory::make_blob((rsl::byte*)&data[0], rsl::memory_size(sizeof(T) * num));
```

Comment:
please create a function overload for memory::make_blob that accepts a T* so you don't have to write this kind of code on call site